Core/tests/Unit/GrpcRequestWrapperTest.php

```diff
@@ -109,6 +109,35 @@ public function responseProvider()
         ];
     }
 
+    public function testRetryFunction()
+    {
+        $this->expectException('Google\Cloud\Core\Exception\GoogleException');
+        $this->expectExceptionMessage('1');
+
+        $requestWrapper = new GrpcRequestWrapper();
+
+        $numCalls = 4;
+        $requestWrapper->send(
+            function () use (&$numCalls) {
+                --$numCalls;
+                throw new ApiException(
+                    $numCalls,
+                    \Google\Rpc\Code::NOT_FOUND,
+                    \Google\ApiCore\ApiStatus::NOT_FOUND
+                );
+            },
+            [[]],
+            [
+                'retries' => $numCalls-2,
+                'retryFunction' => function (\Exception $ex) use (&$numCalls) {
+                    $statusCode = $ex->getCode();
+                    return $statusCode === \Google\Rpc\Code::NOT_FOUND &&
+                        $numCalls > 0;
+                }
+            ]
```

Comment:
This looks like it works, but the logic is a bit hard to follow. What if you just did something like this:
```php
    public function testRetryFunction()
    {
        $requestWrapper = new GrpcRequestWrapper();
        $requestWrapper->send(
            $timesCalled = 0;
            function () use (&$timesCalled) {
                if (1 === $timesCalled++) {
                     // succeed on second try
                     return;
                }
                throw new ApiException('Retryable exception');
            },
            [[]],
            [
                'retries' => 1,
                'retryFunction' => function (\Exception $ex) {
                    return $ex->getMessage() === 'Retryable exception';
                }
            ]
        );

        $this->assertEquals(2, $timesCalled);
    }
```

Reply:
I intended to do hit `retryFunction` multiple times, but it seems to provide little value and test runtime was ~5 seconds. 
Your suggestion reduced it to ~1.5 seconds. Thanks 👍🏼

Reply:
You can do that by just making it `2 === $timesCalled++`, increasing `retries` to `2`, and asserting `$timesCalled` at the bottom to be `3`, but I don't think any of that is necessary.